PolyEngine/Core/Src/Optional.hpp

```diff
@@ -0,0 +1,188 @@
+
+	public:
+		const V&  Value() const& { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return *this->value_storage; }
+		      V&  Value() &      { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return *this->value_storage; }
+
+	public:
+		V& TakeValue() { ASSERTE(this->HasValue(), "Attempting to take a value from an empty optional"); V& ret = *this->Value(); this->value_storage = nullptr; return ret; }
+
+	public:
+		V ValueOr(V&& fallback) const &  { if (this->HasValue()) { return *this->Value(); } return fallback; }
+		V ValueOr(V&& fallback) const && { if (this->HasValue()) { return std::move(*this->Value()); } return fallback; }
+
+	private:
+		V* value_storage;
+	};
+
+	//todo(vuko): this is pretty much copy-pasted from the main implementation (with the copying constructor removed); can we somehow make an alias?
+	//always moves in the contents
+	template<typename V>
+	class Optional<V&&> {
+	public:
+		using ValueType = V&&;
+	public:
+		Optional() : initialized(false), dummy{} {}
+		Optional(      V&& value) : initialized(true), value_storage(std::move(value)) {}
+		~Optional() { if (this->HasValue()) { this->Value().~V(); } }
+
+	public:
+		Optional(Optional&& other) {
+			if (other.HasValue()) {
+				this->initialized = true;
+				::new(&this->Value()) V(std::move(other.Value()));
+
+				other.Value().~V();
+				other.initialized = false;
+			} else {
+				this->initialized = false;
+			}
+		}
+		Optional& operator=(Optional&& other) {
+			if (other.HasValue()) {
+				this->initialized = true;
+				this->Value() = std::move(other.Value());
+
+				other.Value().~V();
+				other.initialized = false;
+			} else if (this->HasValue()) {
+				this->Value().~V();
+				this->initialized = false;
+			}
+			return *this;
+		}
+
+	public:
+		bool HasValue() const { return this->initialized; }
+		operator bool() const { return this->HasValue(); }
+
+	public:
+		const V&  Value() const& { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		V&  Value() &      { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		V&& Value() &&     { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return std::move(this->value_storage); }
+
+	public:
+		V TakeValue() {
+			ASSERTE(this->HasValue(), "Attempting to take a value from an empty optional");
+
+			V ret = std::move(this->value_storage);
+
+			this->initialized = false;
+			this->value_storage.~V();
+
+			return ret;
+		}
+
+	public:
+		V ValueOr(V&& fallback) const &  { if (this->HasValue()) { return this->Value(); } return fallback; }
+		V ValueOr(V&& fallback) const && { if (this->HasValue()) { return std::move(this->Value()); } return fallback; }
+
+	private:
+		bool initialized;
+		union {
+			V value_storage;
+			char dummy;
+		};
+	};
+
+	//todo(vuko): Optional-creating utility functions
```

Comment:
That ugly thing? :P I fail to see how it is useful, when one can simply use `{}`.
```
Optional<T> func_a() {
	if (cond) {
		return {}; //clearly empty
	} else {
		return {value};
	}
}
```
as opposed to:
```
std::optional<T> func_b() {
	if (cond) {
		return std::nullopt; //what is this...? let me see the docs
	} else {
		return {value};
	}
}
```
Emptying an optional can be done by calling `TakeValue()`.